compiler/p/codegen/OMRTreeEvaluator.cpp

```diff
@@ -5649,13 +5650,453 @@ static TR::Register *inlineArrayCmpP10(TR::Node *node, TR::CodeGenerator *cg, bo
+   TR::Register *cr6Reg = cg->allocateRegister(TR_CCR);
+
+   /* On 32bit, the length value will never exceed a 32bit value so only the lower 32 bits are needed. */
+   if (!is64bit)
+      {
+      pairReg = lengthReg;
+      lengthReg = lengthReg->getLowOrder();
+      }
+
+   TR::LabelSymbol *startLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *load16LoopLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *skipLoad16Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *skipLoad8Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *skipLoad4PreZeroCheckLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *skipLoad4Label = generateLabelSymbol(cg);
+   TR::LabelSymbol *load4DiffLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *load8DiffLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *load16DiffLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *returnFalseBeforeLoadLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *returnFalseLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *returnTrueLabel = generateLabelSymbol(cg);
+   TR::LabelSymbol *endLabel = generateLabelSymbol(cg);
+
+   int32_t numRegs = 10;
+   TR::RegisterDependencyConditions *dependencies = new (cg->trHeapMemory()) TR::RegisterDependencyConditions(0, numRegs, cg->trMemory());
+   dependencies->addPostCondition(src1AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(src2AddrReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(lengthReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(offsetReg, TR::RealRegister::NoReg);
+   dependencies->getPostConditions()->getRegisterDependency(3)->setExcludeGPR0();
+   dependencies->addPostCondition(tempReg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(tempReg2, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(vec1Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(vec2Reg, TR::RealRegister::NoReg);
+   dependencies->addPostCondition(cr0Reg, TR::RealRegister::cr0);
+   dependencies->addPostCondition(cr6Reg, TR::RealRegister::cr6);
+
+   generateLabelInstruction(cg, TR::InstOpCode::label, node, startLabel);
+   startLabel->setStartInternalControlFlow();
+
+   generateTrg1ImmInstruction(cg, TR::InstOpCode::li, node, offsetReg, 0);
+
+   /* If array length is less than 4 jump to the byte by byte handling. */
+   generateTrg1Src1ImmInstruction(cg, is64bit ? TR::InstOpCode::cmpli8 : TR::InstOpCode::cmpli4, node, cr6Reg, lengthReg, 4);
+   generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, skipLoad4PreZeroCheckLabel, cr6Reg);
+
+   if (is64bit)
+      {
+      /* Array length is at least 4. If it is less than 8, jump to the section that loads 4 bytes from the array. */
+      generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::cmpli8, node, cr6Reg, lengthReg, 8);
+      generateConditionalBranchInstruction(cg, TR::InstOpCode::blt, node, skipLoad8Label, cr6Reg);
+      }
+
+   /*
```

Comment:
here, you have an issue for choice of laying down optimal sequence and catching the most likely cases fast. for example, length more than 16 is most the likely. doing a single 16 comparison up front, then handling the residual looks better, instead of paying the cost of 2 taken branches (6 cycles in total usually on P, assuming every branch is perfectly predicted). progressively testing for longer lengths might end up paying the highest cost for most situations. have you thought through it?